doc/tutorials/dnn/dnn_custom_layers/dnn_custom_layers.md

```diff
@@ -0,0 +1,373 @@
+# Custom deep learning layers support in OpenCV {#tutorial_dnn_custom_layers}
+
+## Introduction
+Deep learning is a fast growing area. The new approaches to build neural networks
+usually introduce new types of layers. They could be modifications of existing
+ones or implement outstanding researching ideas.
+
+OpenCV gives an opportunity to import and run networks from different deep learning
+frameworks. There are a number of the most popular layers. However you can face
+a problem that your network cannot be imported using OpenCV because of unimplemented layers.
+
+The first solution is to create a feature request at https://github.com/opencv/opencv/issues
+mentioning details such a source of model and type of new layer. A new layer could
+be implemented if OpenCV community shares this need.
+
+The second way is to define a **custom layer** so OpenCV's deep learning engine
+will know how to use it. This tutorial is dedicated to show you a process of deep
+learning models import customization.
+
+## Define a custom layer in C++
+Deep learning layer is a building block of network's pipeline.
+It has connections to **input blobs** and produces results to **output blobs**.
+There are trained **weights** and **hyper-parameters**.
+Layers' names, types, weights and hyper-parameters are stored in files are generated by
+native frameworks during training. If OpenCV mets unknown layer type it throws an
+exception trying to read a model:
+
+```
+Unspecified error: Can't create layer "layer_name" of type "MyType" in function getLayerInstance
+```
+
+To import the model correctly you have to derive a class from cv::dnn::Layer with
+the following methods:
+
+~~~~~~~~~~~~~{.cpp}
+class MyLayer : public cv::dnn::Layer
```

Comment:
I believe it is better to place whole sample code into a separate file (which could be checked by compiler) and put here snippets only.
See [this example](https://github.com/opencv/opencv/blob/3.4.1/samples/cpp/tutorial_code/core/mat_mask_operations/mat_mask_operations.cpp).

Reply:
@alalek, thanks! Done.